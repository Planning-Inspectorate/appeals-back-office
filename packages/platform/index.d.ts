import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
