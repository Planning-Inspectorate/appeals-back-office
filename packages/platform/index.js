export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
