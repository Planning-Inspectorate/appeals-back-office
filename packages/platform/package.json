{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
