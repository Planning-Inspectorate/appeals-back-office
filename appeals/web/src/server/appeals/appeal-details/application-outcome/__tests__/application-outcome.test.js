import { parseHtml } from '@pins/platform';
import supertest from 'supertest';
import { appealData } from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';

describe('application outcome', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	describe('GET /change', () => {
		it('should render application outcome page from appeals details', async () => {
			const appealId = appealData.appealId;
			const response = await request.get(
				`${baseUrl}/${appealId}/appellant-case/application-outcome/change`
			);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('What was the outcome of the application?</h1>');

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain(
				'name="applicationOutcomeRadio" type="radio" value="granted"'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="applicationOutcomeRadio" type="radio" value="refused"'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="applicationOutcomeRadio" type="radio" value="not_received"'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});
	});

	describe('POST /change', () => {
		it('should re-direct to lpaq page when data is valid for LPA and came from lpa-questionnaire', async () => {
			const appealId = appealData.appealId;
			const validData = {
				lpaChangedDescriptionRadio: 'granted'
			};

			const response = await request
				.post(`${baseUrl}/${appealId}/appellant-case/application-outcome`)
				.send(validData);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				`Found. Redirecting to appeals-service/appeal-details/${appealId}/appellant-case/change`
			);
		});
	});
});
