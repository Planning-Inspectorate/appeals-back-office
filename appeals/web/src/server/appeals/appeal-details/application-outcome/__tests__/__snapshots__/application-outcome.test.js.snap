// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application-outcome GET /change shouldrender the applicationOutcome change page 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">What was the application decision?</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="application-outcome" name="application-outcome"
                            type="radio" value="granted">
                            <label class="govuk-label govuk-radios__label" for="application-outcome">Granted</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="application-outcome-2" name="application-outcome"
                            type="radio" value="refused">
                            <label class="govuk-label govuk-radios__label" for="application-outcome-2">Refused</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="application-outcome-3" name="application-outcome"
                            type="radio" value="not_received">
                            <label class="govuk-label govuk-radios__label" for="application-outcome-3">Not received</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`application-outcome POST /change should re-render applicationOutcome change page if data is invalid 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-error-summary" data-module="govuk-error-summary">
                <div role="alert">
                    <h2 class="govuk-error-summary__title"> There is a problem</h2>
                    <div class="govuk-error-summary__body">
                        <ul class="govuk-list govuk-error-summary__list">
                            <li><a href="#application-outcome">Select one of the options</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">What was the application decision?</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="application-outcome" name="application-outcome"
                            type="radio" value="granted">
                            <label class="govuk-label govuk-radios__label" for="application-outcome">Granted</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="application-outcome-2" name="application-outcome"
                            type="radio" value="refused">
                            <label class="govuk-label govuk-radios__label" for="application-outcome-2">Refused</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="application-outcome-3" name="application-outcome"
                            type="radio" value="not_received">
                            <label class="govuk-label govuk-radios__label" for="application-outcome-3">Not received</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;
